// tanstack_start_ts/src/components/credit-card-checkout-form-card.tsx
import * as React from "react";
import type { StardormChatRichBlock } from "@railbeam/stardorm-api-contract";
import {
  ISO_3166_1_ALPHA2_CODES,
  isoCountryDisplayName,
} from "@railbeam/stardorm-api-contract";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import {
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components/ui/select";
import { cn } from "@/lib/utils";
import { toast } from "sonner";

type CreditCardFormRich = Extract<
  StardormChatRichBlock,
  { type: "credit_card_checkout_form" }
>;

const COUNTRY_SELECT_OPTIONS: { code: string; label: string }[] = (() => {
  return [...ISO_3166_1_ALPHA2_CODES]
    .map((code) => {
      const name = isoCountryDisplayName(code);
      return {
        code,
        label: name !== code ? `${name} (${code})` : code,
      };
    })
    .sort((a, b) => a.label.localeCompare(b.label, "en"));
})();

export function CreditCardCheckoutFormCard({
  rich,
  disabled,
  onSubmitCard,
}: {
  rich: CreditCardFormRich;
  disabled?: boolean;
  onSubmitCard: (params: Record<string, unknown>) => void | Promise<void>;
}) {
  const defaultCurrency = (
    rich.defaultCurrency?.trim().toUpperCase() || "USD"
  ).slice(0, 3);

  const [firstName, setFirstName] = React.useState("");
  const [lastName, setLastName] = React.useState("");
  const [line1, setLine1] = React.useState("");
  const [line2, setLine2] = React.useState("");
  const [city, setCity] = React.useState("");
  const [region, setRegion] = React.useState("");
  const [postalCode, setPostalCode] = React.useState("");
  const [countryCode, setCountryCode] = React.useState("US");
  const [cardLabel, setCardLabel] = React.useState("");
  const [currency, setCurrency] = React.useState(defaultCurrency);
  const [openingUsd, setOpeningUsd] = React.useState("");

  React.useEffect(() => {
    const d = (rich.defaultCurrency?.trim().toUpperCase() || "USD").slice(0, 3);
    setCurrency(d);
  }, [rich.defaultCurrency]);

  const onSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    const fn = firstName.trim();
    const ln = lastName.trim();
    const l1 = line1.trim();
    const c = city.trim();
    const reg = region.trim();
    const zip = postalCode.trim();
    let cc = countryCode.trim().toUpperCase();
    if (cc === "UK") cc = "GB";
    if (!/^[A-Z]{2}$/.test(cc) || !ISO_3166_1_ALPHA2_CODES.includes(cc)) {
      toast.error("Pick a valid ISO country");
      return;
    }
    if (!fn || !ln || !l1 || !c || !reg || !zip) {
      toast.error("Fill in every required billing field");
      return;
    }
    const cur = currency.trim().toUpperCase();
    if (!/^[A-Z]{3}$/.test(cur)) {
      toast.error("Currency must be a 3-letter ISO code (e.g. USD)");
      return;
    }
    const params: Record<string, unknown> = {
      firstName: fn,
      lastName: ln,
      line1: l1,
      city: c,
      region: reg,
      postalCode: zip,
      countryCode: cc,
      currency: cur,
    };
    const l2 = line2.trim();
    if (l2) params.line2 = l2;
    const label = cardLabel.trim();
    if (label) params.cardLabel = label;
    const usdRaw = openingUsd.trim();
    if (usdRaw) {
      const n = Number.parseFloat(usdRaw);
      if (!Number.isFinite(n) || n < 0) {
        toast.error("Opening balance must be a non-negative number");
        return;
      }
      params.initialBalanceCents = Math.round(n * 100);
    }
    void onSubmitCard(params);
  };

  return (
    <form
      onSubmit={onSubmit}
      className="w-full max-w-md space-y-3 overflow-hidden rounded-xl border border-border bg-surface-elevated p-3.5"
    >
      <div className="border-b border-border pb-2">
        <h3 className="text-sm font-semibold">{rich.title}</h3>
        {rich.intro ? (
          <p className="mt-1 text-xs text-muted-foreground leading-relaxed">{rich.intro}</p>
        ) : (
          <p className="mt-1 text-xs text-muted-foreground leading-relaxed">
            Enter the legal cardholder name and billing address. Values are sent only when you
            create the card.
          </p>
        )}
      </div>

      <div className="grid grid-cols-2 gap-2">
        <div className="space-y-1">
          <Label htmlFor="cc-first" className="text-xs">
            Legal first name
          </Label>
          <Input
            id="cc-first"
            className="h-9 text-sm"
            value={firstName}
            onChange={(ev) => setFirstName(ev.target.value)}
            autoComplete="given-name"
            required
          />
        </div>
        <div className="space-y-1">
          <Label htmlFor="cc-last" className="text-xs">
            Legal last name
          </Label>
          <Input
            id="cc-last"
            className="h-9 text-sm"
            value={lastName}
            onChange={(ev) => setLastName(ev.target.value)}
            autoComplete="family-name"
            required
          />
        </div>
      </div>

      <div className="space-y-1">
        <Label htmlFor="cc-line1" className="text-xs">
          Street address
        </Label>
        <Input
          id="cc-line1"
          className="h-9 text-sm"
          value={line1}
          onChange={(ev) => setLine1(ev.target.value)}
          autoComplete="address-line1"
          required
        />
      </div>
      <div className="space-y-1">
        <Label htmlFor="cc-line2" className="text-xs">
          Apt / suite (optional)
        </Label>
        <Input
          id="cc-line2"
          className="h-9 text-sm"
          value={line2}
          onChange={(ev) => setLine2(ev.target.value)}
          autoComplete="address-line2"
        />
      </div>

      <div className="grid grid-cols-2 gap-2">
        <div className="space-y-1">
          <Label htmlFor="cc-city" className="text-xs">
            City
          </Label>
          <Input
            id="cc-city"
            className="h-9 text-sm"
            value={city}
            onChange={(ev) => setCity(ev.target.value)}
            autoComplete="address-level2"
            required
          />
        </div>
        <div className="space-y-1">
          <Label htmlFor="cc-region" className="text-xs">
            State / region
          </Label>
          <Input
            id="cc-region"
            className="h-9 text-sm"
            value={region}
            onChange={(ev) => setRegion(ev.target.value)}
            autoComplete="address-level1"
            required
          />
        </div>
      </div>

      <div className="grid grid-cols-2 gap-2">
        <div className="space-y-1">
          <Label htmlFor="cc-postal" className="text-xs">
            Postal code
          </Label>
          <Input
            id="cc-postal"
            className="h-9 text-sm"
            value={postalCode}
            onChange={(ev) => setPostalCode(ev.target.value)}
            autoComplete="postal-code"
            required
          />
        </div>
        <div className="space-y-1">
          <Label className="text-xs">Country</Label>
          <Select value={countryCode} onValueChange={setCountryCode}>
            <SelectTrigger className="h-9 text-left text-xs sm:text-sm">
              <SelectValue placeholder="Country" />
            </SelectTrigger>
            <SelectContent className="max-h-72">
              {COUNTRY_SELECT_OPTIONS.map((o) => (
                <SelectItem key={o.code} value={o.code} className="text-xs sm:text-sm">
                  {o.label}
                </SelectItem>
              ))}
            </SelectContent>
          </Select>
        </div>
      </div>

      <div className="grid grid-cols-2 gap-2">
        <div className="space-y-1">
          <Label htmlFor="cc-cur" className="text-xs">
            Currency (ISO)
          </Label>
          <Input
            id="cc-cur"
            className="h-9 font-mono text-sm uppercase"
            maxLength={3}
            value={currency}
            onChange={(ev) => setCurrency(ev.target.value.toUpperCase())}
            autoComplete="off"
          />
        </div>
        <div className="space-y-1">
          <Label htmlFor="cc-open" className="text-xs">
            Opening balance ({currency || "USD"}, optional)
          </Label>
          <Input
            id="cc-open"
            className="h-9 text-sm"
            inputMode="decimal"
            placeholder="0.00"
            value={openingUsd}
            onChange={(ev) => setOpeningUsd(ev.target.value)}
          />
        </div>
      </div>

      <div className="space-y-1">
        <Label htmlFor="cc-label" className="text-xs">
          Card label (optional)
        </Label>
        <Input
          id="cc-label"
          className="h-9 text-sm"
          value={cardLabel}
          onChange={(ev) => setCardLabel(ev.target.value)}
          placeholder="e.g. Marketing spend"
        />
      </div>

      <Button
        type="submit"
        size="sm"
        variant="secondary"
        className={cn("w-full")}
        loading={disabled}
        disabled={disabled}
      >
        {disabled ? "Creating…" : "Create virtual card"}
      </Button>
    </form>
  );
}
