// tanstack_start_ts/src/routes/docs/payments.tsx
import { createFileRoute, Link } from "@tanstack/react-router";
import { DocCallout } from "@/components/docs/doc-callout";
import { DocCode } from "@/components/docs/doc-code";
import { DocPageHero, DocProse, DocSection } from "@/components/docs/doc-page";
import { DocResult } from "@/components/docs/doc-result";

export const Route = createFileRoute("/docs/payments")({
  component: DocsPayments,
});

function DocsPayments() {
  return (
    <DocProse>
      <DocPageHero
        eyebrow="Checkout"
        title="Payments API"
        description="Public read endpoints for hosted checkouts plus settlement recording after an on-chain transfer. x402 rows are created with the authenticated handler and expose machine-readable accepts metadata."
      />

      <DocSection title="x402 payment (create → checkout → settle)">
        <p className="text-sm text-muted-foreground">
          <a href="https://www.x402.org/" className="text-foreground underline-offset-4 hover:underline">
            x402
          </a>{" "}
          checkouts model paywalled HTTP resources. Use{" "}
          <code className="text-foreground">sdk.handlers.invoke</code> with a signed-in wallet session (same auth as{" "}
          <code className="text-foreground">/users/me/*</code>) to persist a row, then share the hosted{" "}
          <code className="text-foreground">/pay/:id</code> link or drive settlement from your own wallet code.
        </p>
        <DocCode title="Create checkout (handler)">
          {`const created = await sdk.handlers.invoke("create_x402_payment", {
  id: "usage-pack-jan-2026",
  amount: "1000000000000000", // smallest units — wei for native, or base units for ERC-20
  currency: "native", // or \`0x…\` token contract on the target network
  network: "16602", // e.g. 0G testnet chain id string; use your deployment’s id or slug
  payTo: "0x742d35Cc6634C0532925a3b844Bc454e4438f44e",
  title: "API access — starter",
  description: "Unblocks your resource after the payer submits a matching on-chain transfer.",
  resourceUrl: "https://api.example.com/v1/premium/status",
  decimals: 18,
});

const { paymentRequestId, payPath } = created as {
  paymentRequestId: string;
  payPath: string;
};
// Open \`baseUrl + payPath\` for the hosted wallet UI (e.g. "/pay/664b2f…").`}
        </DocCode>
        <DocResult title="Handler payload (abridged)">
          {`{
  "x402Version": 1,
  "resource": {
    "id": "usage-pack-jan-2026",
    "key": "16602:usage-pack-jan-2026",
    "title": "API access — starter",
    "resourceUrl": "https://api.example.com/v1/premium/status"
  },
  "accepts": [
    {
      "scheme": "exact",
      "network": "16602",
      "maxAmountRequired": "1000000000000000",
      "asset": "native",
      "payTo": "0x742d35cc6634c0532925a3b844bc454e4438f44e",
      "resource": "16602:usage-pack-jan-2026",
      "description": "Unblocks your resource after the payer submits a matching on-chain transfer."
    }
  ],
  "paymentRequestId": "664b2f…",
  "payPath": "/pay/664b2f…"
}`}
        </DocResult>
        <DocCode title="Read public checkout + record tx">
          {`const payment = await sdk.payments.get(paymentRequestId);
// payment.type === "x402" — optional facilitator-specific fields in payment.x402Payload

await sdk.payments.pay(payment.id, {
  txHash: "0xabc…",
  payerAddress: wallet.address,
});`}
        </DocCode>
        <DocCallout variant="info" title="Agents vs SDK">
          <p>
            In chat, agents call the same capability as the <code className="text-foreground">create_x402_payment</code>{" "}
            tool; the SDK path is <code className="text-foreground">POST /handlers/create_x402_payment</code> with your
            session headers. Listing runnable ids: <code className="text-foreground">sdk.handlers.list()</code> (see{" "}
            <Link to="/docs/handlers-storage" className="text-foreground underline-offset-4 hover:underline">
              Handlers & storage
            </Link>
            ).
          </p>
        </DocCallout>
      </DocSection>

      <DocSection title="Fetch a checkout row">
        <DocCode title="sdk.payments.get">
          {`const payment = await sdk.payments.get("664b2f…"); // mongo id`}
        </DocCode>
        <DocResult title="PublicPaymentRequest (abridged)">
          {`{
  "id": "664b2f…",
  "type": "on-chain",
  "status": "pending",
  "title": "Agent subscription",
  "amount": "1000000",
  "asset": "native",
  "payTo": "0x…",
  "network": "16602"
}`}
        </DocResult>
        <DocCallout variant="info" title="No implicit list endpoint">
          <p>
            The REST controller exposes <code className="text-foreground">GET /payments/:id</code> for anonymous
            checkout pages. Authenticated lists live under <code className="text-foreground">sdk.users.listPaymentRequests</code>.
          </p>
        </DocCallout>
      </DocSection>

      <DocSection title="Record settlement">
        <p className="text-sm text-muted-foreground">
          For direct wallet settlement, post the EVM transaction hash. When the API is wired to an x402 facilitator, you
          can instead send <code className="text-foreground">x402PaymentPayload</code> (see{" "}
          <code className="text-foreground">paymentSettlementBodySchema</code> in{" "}
          <code className="text-foreground">@railbeam/stardorm-api-contract</code>) — mutually exclusive with{" "}
          <code className="text-foreground">txHash</code>.
        </p>
        <DocCode title="sdk.payments.pay">
          {`await sdk.payments.pay("664b2f…", {
  txHash: "0xabc…",
  payerAddress: wallet.address,
});`}
        </DocCode>
      </DocSection>
    </DocProse>
  );
}
