// tanstack_start_ts/src/routes/settings.tsx
import { createFileRoute } from "@tanstack/react-router";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import * as React from "react";
import { toast } from "sonner";
import type { UpdateUserBody, UserAvatarPreset } from "@railbeam/stardorm-api-contract";
import { useApp } from "@/lib/app-state";
import { fetchStardormMe, patchStardormUser } from "@/lib/stardorm-user-api";
import { getStardormApiBase } from "@/lib/stardorm-axios";
import { queryKeys } from "@/lib/query-keys";
import { useUserAvatarPreset } from "@/lib/hooks/use-user-avatar-preset";
import { USER_AVATAR_URLS } from "@/lib/user-avatar-assets";
import { writeStoredUserAvatarPreset } from "@/lib/user-avatar-preset-storage";
import { cn } from "@/lib/utils";
import { Button } from "@/components/ui/button";
import { Switch } from "@/components/ui/switch";
import { Input } from "@/components/ui/input";
import { Label } from "@/components/ui/label";
import { CoinIcon } from "@/components/icons";
import { PageRoutePending, SettingsProfileFieldsSkeleton } from "@/components/page-shimmer";

export const Route = createFileRoute("/settings")({
  component: Settings,
  pendingComponent: () => <PageRoutePending variant="narrow" />,
});

function Settings() {
  const { address, balance, disconnect, isStardormAuthed } = useApp();
  const queryClient = useQueryClient();
  const userKey = address ? (address.toLowerCase() as `0x${string}`) : null;
  const apiConfigured = Boolean(getStardormApiBase());

  const meQuery = useQuery({
    queryKey: queryKeys.user.me(userKey),
    queryFn: fetchStardormMe,
    enabled: Boolean(apiConfigured && isStardormAuthed && userKey),
  });

  const profileDraftsInit = React.useRef(false);
  const [displayNameDraft, setDisplayNameDraft] = React.useState("");
  const [emailDraft, setEmailDraft] = React.useState("");

  React.useEffect(() => {
    if (!isStardormAuthed) profileDraftsInit.current = false;
  }, [isStardormAuthed]);

  React.useEffect(() => {
    if (meQuery.isSuccess && meQuery.data && !profileDraftsInit.current) {
      setDisplayNameDraft(meQuery.data.displayName ?? "");
      setEmailDraft(meQuery.data.email ?? "");
      profileDraftsInit.current = true;
    }
  }, [meQuery.isSuccess, meQuery.data]);

  const updateMutation = useMutation({
    mutationFn: patchStardormUser,
    onSuccess: (user) => {
      if (userKey) {
        queryClient.setQueryData(queryKeys.user.me(userKey), user);
      }
    },
  });

  const saveProfile = () => {
    updateMutation.mutate(
      {
        displayName: displayNameDraft.trim(),
        email: emailDraft.trim() ? emailDraft.trim().toLowerCase() : null,
      },
      {
        onSuccess: () => toast.success("Profile saved"),
        onError: (e: Error) => {
          toast.error("Could not save profile", { description: e.message });
        },
      },
    );
  };

  const patchPreferences = (preferences: NonNullable<UpdateUserBody["preferences"]>) => {
    updateMutation.mutate(
      { preferences },
      {
        onError: (e: Error) => {
          toast.error("Could not update preference", { description: e.message });
        },
      },
    );
  };

  const prefs = meQuery.data?.preferences;
  const prefsReady = meQuery.isSuccess && Boolean(meQuery.data);
  const syncDisabled = !apiConfigured || !isStardormAuthed || !userKey;
  const prefsBusy = updateMutation.isPending || (isStardormAuthed && meQuery.isPending);

  const avatarPreset = useUserAvatarPreset();

  const setAvatarPreset = (p: UserAvatarPreset) => {
    writeStoredUserAvatarPreset(p);
    if (userKey && meQuery.data) {
      queryClient.setQueryData(queryKeys.user.me(userKey), {
        ...meQuery.data,
        preferences: { ...meQuery.data.preferences, avatarPreset: p },
      });
    }
    if (!syncDisabled && prefsReady) {
      patchPreferences({ avatarPreset: p });
    }
  };

  const avatarChoices = (
    <div className="flex flex-wrap gap-3">
      {(["male", "female"] as const).map((key) => (
        <button
          key={key}
          type="button"
          onClick={() => setAvatarPreset(key)}
          className={cn(
            "relative h-16 w-16 overflow-hidden rounded-full border-2 transition-shadow focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background",
            avatarPreset === key
              ? "border-primary ring-2 ring-primary ring-offset-2 ring-offset-background"
              : "border-transparent hover:border-border",
          )}
          aria-pressed={avatarPreset === key}
          aria-label={key === "male" ? "Male avatar" : "Female avatar"}
        >
          <img src={USER_AVATAR_URLS[key]} alt="" className="h-full w-full object-cover" />
        </button>
      ))}
    </div>
  );

  return (
    <div className="px-4 py-6 md:px-10 md:py-8">
      <div className="mx-auto max-w-3xl">
        <div className="text-sm uppercase tracking-wider text-muted-foreground">
          Settings
        </div>
        <h1 className="text-2xl font-bold md:text-3xl">Account</h1>

        {apiConfigured && !isStardormAuthed && address ? (
          <p className="mt-3 text-sm text-muted-foreground">
            Sign in with your wallet on the Beam API to sync profile and preferences to your account.
          </p>
        ) : null}
        {!apiConfigured ? (
          <p className="mt-3 text-sm text-muted-foreground">
            Beam API URL is not configured; settings stay on this device only.
          </p>
        ) : null}
        {isStardormAuthed && meQuery.isError ? (
          <p className="mt-3 text-sm text-destructive">
            Could not load your account from the server. Try signing out and signing in again.
          </p>
        ) : null}

        <div className="mt-6 grid gap-4">
          <Card title="Wallet">
            <div className="flex items-center justify-between">
              <div>
                <div className="text-sm">{address ?? "Not connected"}</div>
                <div className="text-sm text-muted-foreground">
                  Network · 0G testnet
                </div>
              </div>
              <div className="flex items-center gap-2 rounded-md border border-border bg-surface-elevated px-3 py-1.5 text-sm">
                <CoinIcon className="h-3.5 w-3.5" />{" "}
                {balance.toLocaleString(undefined, { maximumFractionDigits: 2 })} 0G
              </div>
            </div>
            {address && (
              <div className="mt-3">
                <Button variant="outline" size="sm" onClick={disconnect}>
                  Disconnect
                </Button>
              </div>
            )}
          </Card>

          <Card title="Profile">
            <div className="grid gap-3">
              <div className="grid gap-1.5">
                <Label>Avatar</Label>
                {avatarChoices}
                {!apiConfigured || !isStardormAuthed ? (
                  <p className="text-xs text-muted-foreground">
                    Avatar choice is stored on this device until you sign in to sync it to your account.
                  </p>
                ) : null}
              </div>
              {isStardormAuthed && meQuery.isPending ? (
                <SettingsProfileFieldsSkeleton />
              ) : (
                <>
                  <div className="grid gap-1.5">
                    <Label htmlFor="name">Display name</Label>
                    <Input
                      id="name"
                      value={displayNameDraft}
                      onChange={(e) => setDisplayNameDraft(e.target.value)}
                      placeholder="Beam User"
                      disabled={syncDisabled || meQuery.isPending}
                    />
                  </div>
                  <div className="grid gap-1.5">
                    <Label htmlFor="email">Email (optional)</Label>
                    <Input
                      id="email"
                      type="email"
                      value={emailDraft}
                      onChange={(e) => setEmailDraft(e.target.value)}
                      placeholder="you@beam.0g"
                      disabled={syncDisabled || meQuery.isPending}
                    />
                  </div>
                  <div>
                    <Button
                      size="sm"
                      onClick={saveProfile}
                      loading={updateMutation.isPending}
                      disabled={syncDisabled || meQuery.isPending || updateMutation.isPending}
                    >
                      Save profile
                    </Button>
                  </div>
                </>
              )}
            </div>
          </Card>

          <Card title="Preferences">
            <Pref
              title="Auto-route prompts"
              desc="Let Beam pick the best agent for each prompt."
              checked={prefs?.autoRoutePrompts ?? true}
              disabled={syncDisabled || !prefsReady || prefsBusy}
              onCheckedChange={(v) => patchPreferences({ autoRoutePrompts: v })}
            />
            <Pref
              title="Receipt confirmations"
              desc="Keep a confirmation for each agent action on 0G."
              checked={prefs?.onchainReceipts ?? true}
              disabled={syncDisabled || !prefsReady || prefsBusy}
              onCheckedChange={(v) => patchPreferences({ onchainReceipts: v })}
            />
            <Pref
              title="Email notifications"
              desc="Daily digest of agent activity."
              checked={prefs?.emailNotifications ?? false}
              disabled={syncDisabled || !prefsReady || prefsBusy}
              onCheckedChange={(v) => patchPreferences({ emailNotifications: v })}
            />
          </Card>
        </div>
      </div>
    </div>
  );
}

function Card({ title, children }: { title: string; children: React.ReactNode }) {
  return (
    <div className="rounded-xl border border-border bg-surface p-4">
      <div className="text-sm font-semibold">{title}</div>
      <div className="mt-3">{children}</div>
    </div>
  );
}

function Pref({
  title,
  desc,
  checked,
  disabled,
  onCheckedChange,
}: {
  title: string;
  desc: string;
  checked: boolean;
  disabled?: boolean;
  onCheckedChange: (v: boolean) => void;
}) {
  return (
    <div className="flex items-center justify-between border-t border-border py-3 first:border-0 first:pt-0">
      <div>
        <div className="text-sm">{title}</div>
        <div className="text-sm text-muted-foreground">{desc}</div>
      </div>
      <Switch checked={checked} onCheckedChange={onCheckedChange} disabled={disabled} />
    </div>
  );
}
